python/whylogs/experimental/core/metrics/embedding_metric.py

```diff
@@ -0,0 +1,129 @@
+from whylogs.core.metrics.multimetric import MultiMetric
+from whylogs.core.metrics.serializers import serializer
+from whylogs.core.preprocessing import PreprocessedColumn
+from whylogs.core.proto import MetricComponentMessage
+
+# TODO: share these with NLP code
+
+
+def _serialize_ndarray(a: np.ndarray) -> bytes:
+    bio = BytesIO()
+    np.save(bio, a, allow_pickle=False)
+    return bio.getvalue()
+
+
+def _deserialize_ndarray(a: bytes) -> np.ndarray:
+    bio = BytesIO(a)
+    return np.load(bio, allow_pickle=False)
+
+
+class MatrixComponent(MetricComponent[np.ndarray]):
+    #    mtype = np.ndarray
+    type_id = 101
+
+
+@serializer(type_id=101)
+def serialize(value: np.ndarray) -> MetricComponentMessage:
+    return MetricComponentMessage(serialized_bytes=_serialize_ndarray(value))
+
+
+@deserializer(type_id=101)
+def deserialize(msg: MetricComponentMessage) -> np.ndarray:
+    return _deserialize_ndarray(msg.serialized_bytes)
+
+
+class DistanceFunction(Enum):
+    euclidean = euclidean_distances
+    cosine = cosine_distances
+
+
+@dataclass(frozen=True)
+class EmbeddingConfig(MetricConfig):
+    references: np.ndarray  # columns are reference vectors
+    labels: Optional[List[str]] = None
+    distance_fn: DistanceFunction = DistanceFunction.cosine
+
+    def __post_init__(self) -> None:
+        if len(self.references.shape) != 2:
+            raise ValueError("Embedding reference matrix must be 2 dimensional")
+
+        if self.labels:
+            if len(self.labels) == self.references.shape[1]:
+                raise ValueError(
+                    f"Number of labels ({len(self.labels)}) must match number of reference vectors ({self.references.shape[1]})"
+                )
+
+
+@dataclass
+class EmbeddingMetric(MultiMetric):
+    references: MatrixComponent
+    labels: List[str]
+    distance_fn: DistanceFunction
+
+    def __post_init__(self):
+        submetrics = {
+            f"{label}_distance": {
+                "distribution": StandardMetric.distribution.zero(),
+                "counts": StandardMetric.counts.zero(),
+                "types": StandardMetric.types.zero(),
+            }
+            for label in self.labels
+        }
+        submetrics.update(
+            {
+                "closest": {
+                    "frequent_items": StandardMetric.frequent_items.zero(),
+                    "counts": StandardMetric.counts.zero(),
+                    "types": StandardMetric.types.zero(),
+                }
+            }
+        )
+        super().__init__(submetrics)
+
+    @property
+    def namespace(self) -> str:
+        return "embedding"
+
+    def _update_submetrics(self, submetric: str, data: PreprocessedColumn) -> None:
+        for key in self.submetrics[submetric].keys():
+            self.submetrics[submetric][key].columnar_update(data)
+
+    def columnar_update(self, data: PreprocessedColumn) -> OperationResult:
+        X = data.list.objs  # TODO: throw if not 2D
```

Comment:
I got used to avoiding it from the NLP work :)  It should be fine here though.

Reply:
Actually, this doesn't seem to be working with the `_process_scalar_value()` plumbing :(

Reply:
“Scalar” is a little weird for naming the processing a vector, more I meant I would expect the preprocessing would put the vectors into a NumpyView, it can be a TODO to follow up if the list objects works, but it’s also an “everything” else bucket so I wouldn’t throw if not a vector but skip processing it

Reply:
I made a change in `dataset_profile.py` to get `ndarray` into the Numpy view. Let me know what you think of it.

Reply:
its still a surprising callstack, but maybe thats more of an internal detail. Its something we can follow up on, so we don't have to block on this issue, and better that it gets attached to the NumpyView now.